### PR TITLE
On exit terminate editors that aren't finished yet

### PR DESCRIPTION
**Pull request by phdru**

### CODE REVIEW

**Comment by phdru** [comment]
To be honest I'm not sure termination must be that cruel. I'd prefer to have a checkbox in the web interface "Terminate editors on exit". But I'm a Python programmer and don't know much about WebExtension API. I could try to add something to the already existing interface but that could take a lot of time — I'm extremely busy at the job.

File: native/textern.py [hunk 1]
```diff
@@ -64,6 +65,19 @@ def get(self, relfn):
...
+    def add_editor(self, absfn, proc):
+        relfn = os.path.basename(absfn)
+        assert relfn in self._tmpfiles
+        self._editors[absfn] = proc
+
+    def kill_editors(self):
+        """Terminate editors that aren't finished yet"""
+        for proc in self._editors.values():
+            proc.terminate()
+        # Try harder
```
Comment (cassava) [inline native/textern.py:77]: There should be a timeout between terminate and kill. Or maybe not try harder at all?

**Comment by cassava** [comment]
You should reference issue #44 in the commit, which will then link this PR to that issue.
Also, if I look at the other commits in the repository, @jlebon uses the prefix `textern.py:`, which might be better than `feat(native):`

You can fix the commit by doing an interactive rebase (or just commit amend) followed by a force push. Github should show that change here.

File: native/textern.py (hunk 1 above)
Comment (phdru) [reply native/textern.py:77]: A timeout, ok. Adding this to my TODO…

**Comment by phdru** [comment]
> You should reference issue #44 in the commit, which will then link this PR to that issue.

Ok.

> Also, if I look at the other commits in the repository, @jlebon uses the prefix `textern.py:`, which might be better than `feat(native):`

This time I disagree it's better. It's a [conventional commit](https://www.conventionalcommits.org/) format. The scope could be discussed.

File: native/textern.py [hunk 2]
```diff
@@ -70,6 +72,20 @@ def get(self, relfn):
...
+    def add_editor(self, absfn, proc):
+        relfn = os.path.basename(absfn)
+        assert relfn in self._tmpfiles
+        self._editors[absfn] = proc
+
+    def kill_editors(self):
+        """Terminate editors that aren't finished yet"""
+        for proc in self._editors.values():
+            proc.terminate()
+        sleep(1)
```
Comment (phdru) [inline native/textern.py:84]: The timeout also should be configurable…

**Comment by phdru** [comment]
I'm returning to extend this. I downloaded and installed Firefox Developer's Edition, in `about:config` switched `xpinstall.requires.signature` to `false`.

In the source tree I ran `make xpi`. Then tried to install the xpi into Fox. It rejected the xpi with error "The extension is unverified".

What I'm missing?

**Comment by phdru** [comment]
Currently I create a separate checkbox "Kill editors on exit" and an input field to input timeout between retries. But I think I can use `timeout=0` or `timeout<0` as a signal to not to kill editors.

What would be the better interface? Checkbox seems to be more user-friendly. Using `timeout` with two different purpose seems hackish but lowers the number of interface elements.

**Comment by phdru** [comment]
The new code is untested, unfortunately. I still cannot debug the xpi.

**Comment by phdru** [comment]
Last push: fixed `.value` -> `.checked` for the checkbox "Allow to kill editors" in the options. Still not tested.

**Comment by phdru** [comment]
I managed to install `textern.xpi` into a nightly build of `FFox 92.0a1`. Continue debugging…

**Comment by phdru** [comment]
This is the debugged version. Please review.

File: native/textern.py [hunk 3]
```diff
@@ -70,6 +74,26 @@ def get(self, relfn):
...
+    def kill_editors(self):
+        """Terminate editors that aren't finished yet"""
+        if not self.kill_editors_allow or not self._editors:
+            return
+        for proc in self._editors.values():
+            proc.terminate()
+        sleep(self.kill_editors_timeout)
+        # Try harder
+        for proc in self._editors.values():
+            proc.kill()
```
Comment (jlebon) [inline native/textern.py:95]: A bit unsure about this. I think a cleaner approach would be to use `prctl(PR_SET_PDEATHSIG)`.

File: native/textern.py [hunk 4]
```diff
@@ -78,6 +102,7 @@ def main():
         loop.add_reader(sys.stdin.buffer, handle_stdin, tmp_mgr)
         loop.add_reader(ino.fd, handle_inotify_event, ino, tmp_mgr)
         loop.run_forever()
+        tmp_mgr.kill_editors()
```
Comment (jlebon) [inline native/textern.py:105]: Maybe better to move this after the next line?

**Review by jlebon — COMMENTED** [review]
Thanks for working on this! As long as it's disabled by default, the feature at a high-level sounds reasonable.

File: native/textern.py (hunk 3 above)
Comment (phdru) [reply native/textern.py:95]: > A bit unsure about this. I think a cleaner approach would be to use `prctl(PR_SET_PDEATHSIG)`.

I'm not familiar with `prctl`. AFAIU

1) it is not supported by stdlib; requires additional dependencies; all 3rd-party packages I found at PyPI are quite old and don't work with latest Pythons;

2) `prctl(PR_SET_PDEATHSIG)` is about threads, not processes.

File: native/textern.py (hunk 4 above)
Comment (phdru) [reply native/textern.py:105]: > Maybe better to move this after the next line?

That's possible, no problem.

File: native/textern.py (hunk 3 above)
Comment (jlebon) [reply native/textern.py:95]: It should work to prefix the editor command with `setpriv --pdeathsig SIGTERM`.

Comment (phdru) [reply native/textern.py]: > It should work to prefix the editor command with `setpriv --pdeathsig SIGTERM`.

That requires a complete revamping. I'll do it in a different branch and push to a new PR.

Comment (phdru) [reply native/textern.py]: > It should work to prefix the editor command with `setpriv --pdeathsig SIGTERM`.

Doesn't work for me even in a simple command line test. I open a terminal, run `setpriv --pdeathsig SIGTERM gvim`, and close the terminal; `gvim` stays on screen. Shouldn't it be closed? `pkill -TERM gvim` works (`gvim` stopped) so it receives the signal.

**Comment by jlebon** [comment]
Continued in #88. (FWIW, I wouldn't create a separate PR for this. Just updating this one would've been fine. :) )

**Pull request closed without merging**